src/panels/lovelace/components/hui-action-editor.ts

```diff
@@ -0,0 +1,136 @@
+import { html, LitElement, PropertyDeclarations } from "@polymer/lit-element";
+import { TemplateResult } from "lit-html";
+
+import "../../../components/ha-service-picker";
+
+import { HomeAssistant } from "../../../types";
+import { fireEvent, HASSDomEvent } from "../../../common/dom/fire_event";
+import { EditorTarget } from "../editor/types";
+import { ActionConfig } from "../../../data/lovelace";
+
+declare global {
+  // for fire event
+  interface HASSDomEvents {
+    "action-changed": undefined;
+  }
+  // for add event listener
+  interface HTMLElementEventMap {
+    "action-changed": HASSDomEvent<undefined>;
+  }
+}
+
+export class HuiActionEditor extends LitElement {
+  public config?: ActionConfig;
+  public label?: string;
+  protected hass?: HomeAssistant;
+
+  static get properties(): PropertyDeclarations {
+    return { hass: {}, config: {} };
+  }
+
+  get _action(): string {
+    return this.config!.action || "";
+  }
+
+  get _navigation_path(): string {
+    return this.config!.navigation_path || ""; // TODO Why is this only seeing ToggleActionConfig?
+  }
+
+  get _service(): string {
+    return this.config!.service || ""; // TODO Why is this only seeing ToggleActionConfig?
+  }
+
+  get _service_data(): { [key: string]: any } {
+    return {}; // TODO Need advice on how to handle this free-form data
```

Comment:
Either a text area with JSON or we just don't support it in the UI editor.

Reply:
My "solution" does not work and I need sleep :)
Would appreciate someone looking it over. It should really just be like the yaml editor, right? I just didn't have the time to look at how that was done. Will do so tomorrow.